substrate/genesis-liquidity/pallet/src/lib.rs

```diff
@@ -0,0 +1,394 @@
+            origin.into(),
+            coin,
+            u64::try_from(pool_amount).unwrap(),
+            sri_amount,
+            u64::try_from(pool_amount).unwrap(),
+            sri_amount,
+            GENESIS_LIQUIDITY_ACCOUNT.into(),
+          )
+          .unwrap();
+
+          // let everyone know about the event
+          Self::deposit_event(Event::GenesisLiquidityAddedToPool {
+            coin1: Balance { coin, amount: Amount(u64::try_from(pool_amount).unwrap()) },
+            coin2: Balance { coin: Coin::Serai, amount: Amount(sri_amount) },
+          });
+
+          // set liquidity tokens per account
+          let tokens =
+            u128::from(LiquidityTokens::<T>::balance(GENESIS_LIQUIDITY_ACCOUNT.into(), coin).0);
+          let mut total_tokens_this_coin: u128 = 0;
+
+          let accounts = account_values.get(&coin).unwrap();
+          for (i, (acc, acc_value)) in accounts.iter().enumerate() {
+            // give whatever left to the last account not to have rounding errors.
+            let liq_tokens_this_acc = if i == accounts.len() - 1 {
+              tokens - total_tokens_this_coin
+            } else {
+              tokens.saturating_mul(*acc_value) / pool_value
+            };
+
+            total_tokens_this_coin = total_tokens_this_coin.saturating_add(liq_tokens_this_acc);
+
+            LiquidityTokensPerAddress::<T>::set(
+              coin,
+              acc,
+              Some(u64::try_from(liq_tokens_this_acc).unwrap()),
+            );
+          }
+          assert_eq!(tokens, total_tokens_this_coin);
+        }
+        assert_eq!(total_sri_distributed, GENESIS_SRI);
+
+        // we shouldn't have left any coin in genesis account at this moment, including SRI.
+        // All transferred to the pools.
+        for coin in COINS {
+          assert_eq!(Coins::<T>::balance(GENESIS_LIQUIDITY_ACCOUNT.into(), coin), Amount(0));
+        }
+      }
+
+      // we accept we reached economic security once we can mint smallest amount of a network's coin
+      for coin in COINS {
+        let existing = EconomicSecurityReached::<T>::get(coin.network());
+        if existing == 0u32.into() &&
+          <T as CoinsConfig>::AllowMint::is_allowed(&Balance { coin, amount: Amount(1) })
+        {
+          EconomicSecurityReached::<T>::set(coin.network(), n);
+          Self::deposit_event(Event::EconomicSecurityReached { network: coin.network() });
+        }
+      }
+    }
+  }
+
+  impl<T: Config> Pallet<T> {
+    /// Add genesis liquidity for the given account. All accounts that provide liquidity
+    /// will receive the genesis SRI according to their liquidity ratio.
+    pub fn add_coin_liquidity(account: PublicKey, balance: Balance) -> DispatchResult {
+      // check we are still in genesis period
+      if Self::genesis_ended() {
+        Err(Error::<T>::GenesisPeriodEnded)?;
+      }
+
+      // mint the coins
+      Coins::<T>::mint(GENESIS_LIQUIDITY_ACCOUNT.into(), balance)?;
```

Comment:
Prior mints were done in the II pallet. We may want to keep that pattern?

Reply:
I don't have a strong opinion on this but I would rather keep it here? The prior check is here. Also the reason previous mints were there was because mint account belongs there. Also prob because we didn't want to further modify(adding extra functions etc.) Dex pallet? Again can move it to there with the check if that makes more sense.

Reply:
The argument for it is the coins were flagged as incoming with the II, and the II is just executing upon the incoming coins.

The risk is we mint in II _and_ mint in some module.

I do think II should be the minter. If it mints the coins, and then we fail to execute the II (add genesis liq), it'll get scheduled to get sent back out and not be an issue.